activestorage/lib/active_storage/delivery_method.rb

```diff
@@ -0,0 +1,29 @@
+# frozen_string_literal: true
+
+module ActiveStorage
+  class DeliveryMethod
+    class << self
+      def representation_url(representation, url_options: nil, delivery_method: nil)
+        raise NotImplementedError
+      end
+
+      def blob_url(signed_id, filename, url_options: nil, delivery_method: nil)
+        raise NotImplementedError
+      end
+
+      def select(delivery_method)
+        ActiveStorage.delivery_methods.find do |dm|
```

Comment:
Do we need to keep the list of the available delivery methods? What about inferring a class from the name like we do for other adapters in Rails (e.g., in Active Job  and Action Cable).

And it would be make sense to add a memoization here. Or enforce registering adapters beforehand and keep a mapping, smth like:

```ruby
# this is a shortcut for accessing the mapping
def select(delivery_method)
  ActiveStorage.delivery_methods.fetch(delivery_method)
end

# in config file
config.active_storage.register_delivery_method :redirect

# which translates to
module ActiveStorage
  def self.register_delivery_method(name, options = {})
    delivery_methods[name] = "ActiveStorage::DeliveryMethod::#{name.to_s.classify}".constantize.new(options)
  end 
end
```

Note that I'm using class instances instead of modules—custom delivery methods might have some kind of _state_, e.g.:

```ruby
config.active_storage.register_delivery_method :imgproxy, proxy_url: "https://my.url"
```

That could be useful even for the default `proxy` delivery method: this way we will be able to use different CDNs for different attachments:

```ruby
has_one_attached :avatar, delivery_method: :proxy, deliver_method_options: {url: MY_MAIN_CDN}

has_one_attached :video, delivery_method: :proxy, deliver_method_options: {url: MY_VIDEO_CDN}
```

Another related thought: shouldn't we decouple delivery method ids from implementations? I mean, not to specify the exact method (`redirect` or `proxy`) but the app specific name (e.g., `content_delivery` / `image_delivery`), and then keep a mapping of `ID -> delivery_method`, and use IDs in the app:

```ruby
has_one_attached :avatar, delivery_method: :image_delivery

app.config.active_storage.delivery_method = :content_delivery
```

The same way we deal with a storage backend (via `storage.yml`).